Update Zinc used by the Scala plugin to 1.6.1

tgodzik opened the pull request:
<!--- The issue this PR addresses -->
Fixes https://github.com/gradle/gradle/issues/15491

### Context

The currently used version of zinc is havily outdated and missing a lot of new fixes.  It's also being depracated and will most likely stop working at some point with Scala 3


### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

tgodzik commented on gradle/verification-keyring.keys:
```diff
@@ -14893,3 +14893,107 @@ i8oRVy9MSV3HFJxF7JpNoeEgRBGWCIQJPZ7UZszl4VxSwDkvGoFpZr2MIcXnm++n
 =FL9o
 -----END PGP PUBLIC KEY BLOCK-----
+
+pub rsa2048/cb3808e03cd602a6dc5b882552931f4b72b4f54c 2014-07-19T14:24:53Z
```
These are copied over from https://github.com/gradle/gradle/pull/18003/files

I am not 100% sure what is the proper way of adding these keys and not even sure how to find them.

ljacomet replied:
This should be edited by following the instructions in `gradle/verification-metadata.xml`

tgodzik commented on subprojects/scala/src/integTest/groovy/org/gradle/scala/compile/UpToDateScalaCompileIntegrationTest.groovy:
```diff
@@ -54,8 +54,8 @@ class UpToDateScalaCompileIntegrationTest extends AbstractIntegrationSpec {
         newScalaVersion | newZincVersion
-        '2.11.12'       | '1.2.0'
-        '2.12.6'        | '1.2.5'
+        '2.11.12'       | '1.6.0'
```
It doesn't seem possible to use older Zinc versions. I don't even think this should be possible.

ljacomet replied:
If I understand correctly, this backward compatibility test is broken by your changes.
Compiling Scala 2.11 code is no longer possible, any other recent version _must_ upgrade zinc to 1.6.

In that case, this means this PR should wait until Gradle 8.0 because of that.

Am I right or is there some element I do not understand since I am not a Scala dev?

tgodzik replied:
> Compiling Scala 2.11 code is no longer possible, any other recent version must upgrade zinc to 1.6.

The test still passes, it's just not possible to set Zinc versions to be lower than 1.6.0, since we get binary compat issues (the method invoked is different between versions). However, to be honest it should never be needed to set the Zinc version manually to lower one, since you would be omitting a lot of bugfixes etc. without any possible gain. It's true though if someone does have it set manually it will not work anymore. It's probably possible to make it work via some kind of reflection, but that would require a lot more effort.

The other option would be to deprecate the zinc version setting and ignore it in the settings.

ijuma replied:
@ljacomet why do we say that compiling with Scala 2.11 is not possible?

ljacomet replied:
I do not know - re-reading my comment and I am not sure what I meant exactly 😅 

What remains is the potential backwards compatibility issue. But I also understand that moving forward would be really beneficial to Scala users.
The window for getting this in Gradle 7.5 is still open. Let me check internally if we are OK with this.

tgodzik replied:
The only problem would be if users had the zinc version specified manually, though this should be a really rare scenario and something that should be discouraged. I is a potential compatibility issue though if someone had an old zinc version set.

tgodzik commented:
@ljacomet could you take a look at it?

dubinsky reviewed (approved):
Thanks!!!

ijuma commented on subprojects/scala/build.gradle.kts:
```diff
@@ -29,8 +29,10 @@ dependencies {
 
-    compileOnly("org.scala-sbt:zinc_2.12:1.3.5") {
-        exclude(module="log4j-core") // Because not needed and vulnerable
+    compileOnly("org.scala-sbt:zinc_2.12:1.6.1") {
```
Have we considered using the Scala 2.13 artifact instead?

tgodzik replied:
I haven't thought of that. I can update it for sure, though not sure if benefits will enormous, but I guess we would be fully up to date then.

ijuma replied:
I think the main benefit would be avoiding the Scala 2.12 dependencies given that most projects should be using 2.13 by now.

tgodzik replied:
Changed to 2.13, let's see if everything passes.

ljacomet commented on gradle/verification-metadata.xml:
```diff
@@ -881,6 +883,11 @@
             <pgp value="5b131e826582cf79510daa11cd3e539f208832d0"/>
          </artifact>
       </component>
+      <component group="net.openhft" name="zero-allocation-hashing" version="0.10.1">
```
This entry is not at the right position. Follow instructions in `gradle/verification-metadata.xml` to obtain a stable output.

tgodzik replied:
Now I finally read the README there :sweat_smile: Thanks! It should be correct now.

ljacomet reviewed (changes requested):
Thanks a lot for helping out Gradle Scala users!

See below for a couple comments on dependency verification.

Most important though is the compatibility conversation. I believe we had one already when Zinc 1.5 was released, but I cannot find it ...

dubinsky commented:
@tgodzik Any updates on this? Thanks!

ljacomet reviewed (commented):
First of all, thanks for the contribution.
And sorry it took so long to get a decision for having this in a minor.

But the answer is now yes, so let's see about getting it in Gradle 7.5.


Would you be able to also do some documentation changes:
* Update the [Zinc compatibility table](https://docs.gradle.org/current/userguide/scala_plugin.html#sec:configure_zinc_compiler)
* Add an entry into the potential breaking changes of the upgrade guide - file `upgrading_version_7.adoc`

ljacomet commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"7647598189099c2bcd33db16105ff60f34a05732","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=50926895)

tgodzik commented:
> Looks like running all tests on CI raises a number of problems. Can you reproduce the issues in `org.gradle.scala.compile.*` tests?

I will take a look as soon as possible, just got back from vacation so that might take a bit, but hopefully I should get it fixed by next week.

tgodzik commented:
> Looks like running all tests on CI raises a number of problems. Can you reproduce the issues in `org.gradle.scala.compile.*` tests?

@ljacomet The tests should be fixed, I added a section for the Zinc compatibility table and added a section for potential breaking changes

ljacomet commented:
@bot-gradle test

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"b1d0193dad5ff0ac72e073637a902383cb1bd2e5","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:

ljacomet commented:
@bot-gradle test this

tgodzik commented:
@ljacomet I haven't noticed that some of the tests were failing, but they should be fixed now, protobuf was dropped from zinc (shaded version exists). It is still in Scala 3, but that is most likely related to the fact that the bridge resides in the compiler itself instead of Zinc,

ljacomet commented:
@tgodzik Could you rebase this against the `release` branch?
I would like to try to get it in 7.5.

You will then need to change the target branch to `release` as well.

ljacomet commented:
I have done the rebase and change of base branch.

ljacomet commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"6c8dc28d6561a1aecf4733abf73a4011a5367273","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=51505415)

ljacomet commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":1109499431,"teamCityBuildId":null,"teamCityBuildHeadRef":"584f78e5843d70cbfe53480d62c34928542873a0","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
Sorry some internal error occurs, please contact the administrator @blindpirate

ljacomet commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"584f78e5843d70cbfe53480d62c34928542873a0","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=51507554)

ljacomet commented:
@bot-gradle test this

ljacomet commented:
@bot-gradle test this

tgodzik commented:
> I have done the rebase and change of base branch.

Thanks, sorry that I didn't get to it right away!

ljacomet commented:
No worry, my ability to help steer this was quite spotty 😉

ljacomet reviewed (approved):
Green build!

Thanks @tgodzik for proposing this and following up even after time passed!

ljacomet commented:
@bot-gradle test and merge

bot-gradle commented:
OK, I've already triggered [a build](https://builds.gradle.org/viewQueued.html?itemId=51513325) for you.

mernst-github commented on subprojects/scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompiler.java:
```diff
@@ -197,29 +211,35 @@ public DefinesClass definesClass(File classpathEntry) {
-                if (!e.getKey().exists() || !e.getValue().equals(Stamper.forLastModified().apply(e.getKey()))) {
+            for (Map.Entry<VirtualFileRef, Stamp> e : previousAnalysis.readStamps().getAllLibraryStamps().entrySet()) {
+                File path = CONVERTER.toPath(e.getKey()).toFile();
+                if (!path.exists() || !e.getValue().equals(Stamper.forLastModifiedInRootPaths(CONVERTER).apply(e.getKey()))) {
```
Would like to raise attention to this line which I think is incorrect and seems to have caused https://github.com/gradle/gradle/issues/20101
e.getValue seems to be a `FarmHash`, not a `LastModified`.

mernst-github replied:
TBH I'm not even sure why Gradle provides the `ExternalBinariesLookup` when it effectively could rely on zinc's change detection in the getOrElse case here: https://github.com/sbt/zinc/blob/develop/internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala#L407
